### PR TITLE
update command bus to use only two events

### PR DESCRIPTION
**Pull request by prolic**

### CODE REVIEW

**Comment by prolic** [comment]
ping @ocramius

your thoughts on this? keep working on it or reject and leave as is?

**Comment by codeliner** [comment]
@prolic looks good, more like a pipe now with more or less the same possibilities as before, except that you cannot stop a specific dispatch step but only the entire dispatch. However, I don't think this features is used somewhere at least not in the core. So :+1: for continuing the work

File: src/MessageBus.php [hunk 1]
```diff
@@ -61,75 +61,49 @@ public function deactivate(ActionEventListenerAggregate $plugin): void
...
-        if ($actionEvent->propagationIsStopped()) {
-            throw new RuntimeException('Dispatch has stopped unexpectedly.');
-        }
-    }
-
-    protected function triggerError(ActionEvent $actionEvent): void
-    {
-        $actionEvent->setName(self::EVENT_HANDLE_ERROR);
-
-        $this->getActionEventEmitter()->dispatch($actionEvent);
```
Comment (Ocramius) [inline src/MessageBus.php]: What is the new way of dealing with exceptional scenarios? Just catch?

**Comment by Ocramius** [comment]
IMO this looks good.

Few notes:

 * advantages
    * better performance
    * single event trigger call: can trace everything from there
    * ordering of events is now explicit in the priority, rather than a hidden knowledge
 * disadvantages
    * lost some detail around event naming, since it's now a priority thing

TODOs:

 * [ ] document BC breaks
 * [ ] document migration path
 * [ ] (eventually) benchmark

File: src/MessageBus.php (hunk 1 above)
Comment (prolic) [inline src/MessageBus.php]: From the outside there doesn't change anything. It's still like

```
try {
    $commandBus->dispatch($command);
} catch (DispatchException $e) {
    echo $e->getPrevious()->getMessage();
}
```

If you want to have a listener on it, things changed a little:

```
$commandBus->getActionEventEmitter()->attachListener(
    CommandBus::EVENT_FINALIZE,
    function (ActionEvent $actionEvent) {
        if ($ex = $actionEvent->getParam(CommandBus::EVENT_PARAM_EXCEPTION) {
            // do something
        }
    }
```

**Comment by prolic** [comment]
@codeliner ready for review

documentation will come in seperate PR

**Comment by prolic** [comment]
A litte benchmark

old:
```
<?php

require 'vendor/autoload.php';

$commandBus = new Prooph\ServiceBus\CommandBus();

$commandBus->getActionEventEmitter()->attachListener(
    \Prooph\ServiceBus\CommandBus::EVENT_ROUTE,
    function (\Prooph\Common\Event\ActionEvent $actionEvent): void {
        $actionEvent->setParam(
            \Prooph\ServiceBus\CommandBus::EVENT_PARAM_MESSAGE_HANDLER,
            function (\ProophTest\ServiceBus\Mock\DoSomething $doSomething): void {
            }
        );
    }
);

$command = new \ProophTest\ServiceBus\Mock\DoSomething(['do' => 'it']);

$start = microtime(true);

for ($i = 0; $i < 100000; $i++) {
    $commandBus->dispatch($command);
}

$end = microtime(true);

var_dump((string) ($end - $start) . ' secs for 100000 commands = ' . (string) (($end - $start) / 100000) . ' commands/sec');
```

result: 0.63527607917786 secs for 100000 commands = 6.3527607917786E-6 commands/sec

new:
```
<?php

require 'vendor/autoload.php';

$commandBus = new Prooph\ServiceBus\CommandBus();

$commandBus->getActionEventEmitter()->attachListener(
    \Prooph\ServiceBus\CommandBus::EVENT_DISPATCH,
    function (\Prooph\Common\Event\ActionEvent $actionEvent): void {
        $actionEvent->setParam(
            \Prooph\ServiceBus\CommandBus::EVENT_PARAM_MESSAGE_HANDLER,
            function (\ProophTest\ServiceBus\Mock\DoSomething $doSomething): void {
            }
        );
    },
    \Prooph\ServiceBus\CommandBus::PRIORITY_ROUTE
);

$command = new \ProophTest\ServiceBus\Mock\DoSomething(['do' => 'it']);

$start = microtime(true);

for ($i = 0; $i < 100000; $i++) {
    $commandBus->dispatch($command);
}

$end = microtime(true);

var_dump((string) ($end - $start) . ' secs for 100000 commands = ' . (string) (($end - $start) / 100000) . ' commands/sec');
```

result: 0.55868315696716 secs for 100000 commands = 5.5868315696716E-6 commands/sec

That means 13.7% increased performance.

File: src/EventBus.php [hunk 2]
```diff
@@ -22,53 +22,46 @@ class EventBus extends MessageBus
...
-                $event = $actionEvent->getParam(self::EVENT_PARAM_MESSAGE);
-                $eventListener($event);
-                $actionEvent->setParam(self::EVENT_PARAM_MESSAGE_HANDLED, true);
-            }
-        });
+        $actionEventEmitter->attachListener(
+            self::EVENT_DISPATCH,
+            function (ActionEvent $actionEvent): void {
+                foreach ($actionEvent->getParam(self::EVENT_PARAM_EVENT_LISTENERS, []) as $eventListener) {
+                    if (is_callable($eventListener)) {
```
Comment (Ocramius) [inline src/EventBus.php]: What if it's not callable? Seems weird to not have callables here...

Comment (Ocramius) [inline src/EventBus.php]: Also, can be simplified to

```php
foreach (array_filter($actionEvent->getParam(self::EVENT_PARAM_EVENT_LISTENERS, []), 'is_callable') as $eventListener) {
```

File: src/EventBus.php [hunk 3]
```diff
@@ -22,53 +22,46 @@ class EventBus extends MessageBus
...
+                        $event = $actionEvent->getParam(self::EVENT_PARAM_MESSAGE);
```
Comment (Ocramius) [inline src/EventBus.php]: The event will not change over time - doesn't need to be fetched in the loop

File: src/EventBus.php [hunk 4]
```diff
@@ -22,53 +22,46 @@ class EventBus extends MessageBus
...
+                        $eventListener($event);
+                        $actionEvent->setParam(self::EVENT_PARAM_MESSAGE_HANDLED, true);
```
Comment (Ocramius) [inline src/EventBus.php]: Can be done outside the loop

File: src/MessageBus.php [hunk 5]
```diff
@@ -61,75 +61,49 @@ public function deactivate(ActionEventListenerAggregate $plugin): void
...
+                }
+            },
+            self::PRIORITY_INITIALIZE
+        );
+
+        $actionEventEmitter->attachListener(
+            self::EVENT_DISPATCH,
+            function (ActionEvent $actionEvent): void {
+                if ($actionEvent->getParam(self::EVENT_PARAM_MESSAGE_NAME) === null) {
+                    $message = $actionEvent->getParam(self::EVENT_PARAM_MESSAGE);
```
Comment (Ocramius) [inline src/MessageBus.php]: This variable assignment seems redundant

File: src/EventBus.php (hunk 3 above)
Comment (prolic) [inline src/EventBus.php]: good catch !

File: src/EventBus.php (hunk 4 above)
Comment (prolic) [inline src/EventBus.php]: Good catch!

File: src/MessageBus.php (hunk 5 above)
Comment (prolic) [inline src/MessageBus.php]: done.

**Comment by Ocramius** [comment]
👍

**Pull request merged**